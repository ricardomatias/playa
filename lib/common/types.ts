export enum BinaryEvent {
	Rest = 0,
	Hit = 1
}

export type Octaves = [number, number]

export enum RhythmType {
	Free = 'Free',
	Turn = 'Turn'
}

export type DistributionFunction = <T>(distribution: T | T[], precision?: number | undefined) => string[];
