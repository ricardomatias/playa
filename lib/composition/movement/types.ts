import { Interval, NoteSymbol, ScaleName, TurnMoves } from '../../constants';
import { GreekModeIntervals, ModePosition } from '../../constants/modes';

export const EUCLIDEAN = 'euclid';
export const TURN = 'turn';

export enum MovementRhythm {
	Euclidean = 'Euclidean',
	Turn = 'Turn'
}

export interface TimelineEventKey { root: NoteSymbol, scale: GreekModeIntervals }

export interface TimelineEvent {
	time: number;
	dur: number;
	key: TimelineEventKey;
}

export enum ModulationEventType {
	Mode = 'mode',
	Key = 'key'
}

export interface ModulationEvent {
	key: NoteSymbol;
	root: NoteSymbol;
	type: ModulationEventType;
	scaleName: ScaleName;
	position: ModePosition;
}

export type Movement = {
	timeline: TimelineEvent[];
	events: ModulationEvent[];
}

export interface Turn {
	type: TurnMoves,
	interval: typeof TurnMoves.Free | Interval | number | string
}

/**
 * @typedef {Object} Turn
 * @property {String} type '▼'
 * @property {String | Number} interval '⟷'
 * @example
 * { type: '▼', interval: '⟷' }
*/
export interface TurnEvent extends ModulationEvent {
	turn: Turn
}

export type TurnMovement = {
	timeline: TimelineEvent[];
	events: TurnEvent[];
}
