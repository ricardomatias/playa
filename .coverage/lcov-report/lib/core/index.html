<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for lib/core</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../index.html">All files</a> lib/core
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">88.17% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>343/389</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">81.98% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>141/172</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">83.82% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>57/68</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">87.83% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>332/378</span>
      </div>
    </div>
    <p class="quiet">
      Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
    </p>
  </div>
  <div class='status-line high'></div>
<div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file high" data-value="Chord.js"><a href="Chord.js.html">Chord.js</a></td>
	<td data-value="96.9" class="pic high"><div class="chart"><div class="cover-fill" style="width: 96%;"></div><div class="cover-empty" style="width:4%;"></div></div></td>
	<td data-value="96.9" class="pct high">96.9%</td>
	<td data-value="129" class="abs high">125/129</td>
	<td data-value="90.91" class="pct high">90.91%</td>
	<td data-value="44" class="abs high">40/44</td>
	<td data-value="94.44" class="pct high">94.44%</td>
	<td data-value="18" class="abs high">17/18</td>
	<td data-value="96.77" class="pct high">96.77%</td>
	<td data-value="124" class="abs high">120/124</td>
	</tr>

<tr>
	<td class="file high" data-value="HarmonyBase.js"><a href="HarmonyBase.js.html">HarmonyBase.js</a></td>
	<td data-value="84.62" class="pic high"><div class="chart"><div class="cover-fill" style="width: 84%;"></div><div class="cover-empty" style="width:16%;"></div></div></td>
	<td data-value="84.62" class="pct high">84.62%</td>
	<td data-value="13" class="abs high">11/13</td>
	<td data-value="66.67" class="pct medium">66.67%</td>
	<td data-value="6" class="abs medium">4/6</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	<td data-value="84.62" class="pct high">84.62%</td>
	<td data-value="13" class="abs high">11/13</td>
	</tr>

<tr>
	<td class="file high" data-value="Key.js"><a href="Key.js.html">Key.js</a></td>
	<td data-value="94.67" class="pic high"><div class="chart"><div class="cover-fill" style="width: 94%;"></div><div class="cover-empty" style="width:6%;"></div></div></td>
	<td data-value="94.67" class="pct high">94.67%</td>
	<td data-value="75" class="abs high">71/75</td>
	<td data-value="83.33" class="pct high">83.33%</td>
	<td data-value="36" class="abs high">30/36</td>
	<td data-value="92.31" class="pct high">92.31%</td>
	<td data-value="13" class="abs high">12/13</td>
	<td data-value="94.67" class="pct high">94.67%</td>
	<td data-value="75" class="abs high">71/75</td>
	</tr>

<tr>
	<td class="file low" data-value="Mode.js"><a href="Mode.js.html">Mode.js</a></td>
	<td data-value="0" class="pic low"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="24" class="abs low">0/24</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="9" class="abs low">0/9</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="4" class="abs low">0/4</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="24" class="abs low">0/24</td>
	</tr>

<tr>
	<td class="file high" data-value="Note.js"><a href="Note.js.html">Note.js</a></td>
	<td data-value="89.74" class="pic high"><div class="chart"><div class="cover-fill" style="width: 89%;"></div><div class="cover-empty" style="width:11%;"></div></div></td>
	<td data-value="89.74" class="pct high">89.74%</td>
	<td data-value="78" class="abs high">70/78</td>
	<td data-value="90.48" class="pct high">90.48%</td>
	<td data-value="42" class="abs high">38/42</td>
	<td data-value="84.21" class="pct high">84.21%</td>
	<td data-value="19" class="abs high">16/19</td>
	<td data-value="89.61" class="pct high">89.61%</td>
	<td data-value="77" class="abs high">69/77</td>
	</tr>

<tr>
	<td class="file high" data-value="Scale.js"><a href="Scale.js.html">Scale.js</a></td>
	<td data-value="93.94" class="pic high"><div class="chart"><div class="cover-fill" style="width: 93%;"></div><div class="cover-empty" style="width:7%;"></div></div></td>
	<td data-value="93.94" class="pct high">93.94%</td>
	<td data-value="66" class="abs high">62/66</td>
	<td data-value="82.86" class="pct high">82.86%</td>
	<td data-value="35" class="abs high">29/35</td>
	<td data-value="80" class="pct high">80%</td>
	<td data-value="10" class="abs high">8/10</td>
	<td data-value="93.44" class="pct high">93.44%</td>
	<td data-value="61" class="abs high">57/61</td>
	</tr>

<tr>
	<td class="file empty" data-value="index.js"><a href="index.js.html">index.js</a></td>
	<td data-value="0" class="pic empty"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	</tr>

<tr>
	<td class="file high" data-value="types.js"><a href="types.js.html">types.js</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	</tr>

</tbody>
</table>
</div><div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Wed Feb 20 2019 15:09:45 GMT+0100 (Central European Standard Time)
</div>
</div>
<script src="../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../sorter.js"></script>
<script src="../../block-navigation.js"></script>
</body>
</html>
