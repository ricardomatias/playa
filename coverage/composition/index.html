
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for composition</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../prettify.css" />
    <link rel="stylesheet" href="../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../index.html">All files</a> composition</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">90.49% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>400/442</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">88.29% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>83/94</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">84% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>21/25</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">90.49% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>400/442</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
        <template id="filterTemplate">
            <div class="quiet">
                Filter:
                <input type="search" id="fileSearch">
            </div>
        </template>
    </div>
    <div class='status-line high'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file high" data-value="arp.ts"><a href="arp.ts.html">arp.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="42" class="abs high">42/42</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="8" class="abs high">8/8</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="42" class="abs high">42/42</td>
	</tr>

<tr>
	<td class="file high" data-value="chord-progression.ts"><a href="chord-progression.ts.html">chord-progression.ts</a></td>
	<td data-value="87.69" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 87%"></div><div class="cover-empty" style="width: 13%"></div></div>
	</td>
	<td data-value="87.69" class="pct high">87.69%</td>
	<td data-value="130" class="abs high">114/130</td>
	<td data-value="78.94" class="pct medium">78.94%</td>
	<td data-value="19" class="abs medium">15/19</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="87.69" class="pct high">87.69%</td>
	<td data-value="130" class="abs high">114/130</td>
	</tr>

<tr>
	<td class="file high" data-value="index.ts"><a href="index.ts.html">index.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	</tr>

<tr>
	<td class="file high" data-value="melodies.ts"><a href="melodies.ts.html">melodies.ts</a></td>
	<td data-value="91.56" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 91%"></div><div class="cover-empty" style="width: 9%"></div></div>
	</td>
	<td data-value="91.56" class="pct high">91.56%</td>
	<td data-value="83" class="abs high">76/83</td>
	<td data-value="94.11" class="pct high">94.11%</td>
	<td data-value="17" class="abs high">16/17</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="91.56" class="pct high">91.56%</td>
	<td data-value="83" class="abs high">76/83</td>
	</tr>

<tr>
	<td class="file high" data-value="motif.ts"><a href="motif.ts.html">motif.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="19" class="abs high">19/19</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="5" class="abs high">5/5</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="19" class="abs high">19/19</td>
	</tr>

<tr>
	<td class="file high" data-value="percussion.ts"><a href="percussion.ts.html">percussion.ts</a></td>
	<td data-value="96.42" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 96%"></div><div class="cover-empty" style="width: 4%"></div></div>
	</td>
	<td data-value="96.42" class="pct high">96.42%</td>
	<td data-value="56" class="abs high">54/56</td>
	<td data-value="78.57" class="pct medium">78.57%</td>
	<td data-value="14" class="abs medium">11/14</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="3" class="abs high">3/3</td>
	<td data-value="96.42" class="pct high">96.42%</td>
	<td data-value="56" class="abs high">54/56</td>
	</tr>

<tr>
	<td class="file high" data-value="sequence.ts"><a href="sequence.ts.html">sequence.ts</a></td>
	<td data-value="84.68" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 84%"></div><div class="cover-empty" style="width: 16%"></div></div>
	</td>
	<td data-value="84.68" class="pct high">84.68%</td>
	<td data-value="111" class="abs high">94/111</td>
	<td data-value="90.32" class="pct high">90.32%</td>
	<td data-value="31" class="abs high">28/31</td>
	<td data-value="75" class="pct medium">75%</td>
	<td data-value="16" class="abs medium">12/16</td>
	<td data-value="84.68" class="pct high">84.68%</td>
	<td data-value="111" class="abs high">94/111</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank" rel="noopener noreferrer">istanbul</a>
                at 2024-10-16T10:39:20.013Z
            </div>
        <script src="../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../sorter.js"></script>
        <script src="../block-navigation.js"></script>
    </body>
</html>
    