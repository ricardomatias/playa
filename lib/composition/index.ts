export { createMotif } from './motif';
export { createChordProgression } from './chord-progression';
export { createMelodies } from './melodies';
export { createPercussion } from './percussion';
export { createArp, genHarmonicShift } from './arp';

export * as Rhythm from './rhythm';
export * from './sequence';
export * from './movement';
