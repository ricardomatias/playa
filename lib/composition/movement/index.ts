export { default as createMovement } from './default';
export { default as createTurnMovement } from './turn';
export * from './types';
