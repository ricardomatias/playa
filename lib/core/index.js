/**
* @namespace {module} Core
*/
export { Note } from './Note';
export { default as Chord } from './Chord';
export { default as Scale } from './Scale';
export { default as Key } from './Key';
export { Time } from './Time';
export { Phrase } from './Phrase';
export { PianoRoll } from './PianoRoll';

