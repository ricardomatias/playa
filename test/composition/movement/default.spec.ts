import { Key, Time } from '../../../lib/core';
import { createMovement } from '../../../lib/composition';
import { MovementRhythm } from '../../../lib/composition/movement/types';
import Random from '../../../lib/tools/random';


describe('A Movement test suite', () => {
	it('should generate a basic movement', () => {
		// given
		Random.setSeed('test');

		const dMin = new Key('D', Key.Minor);
		// when
		const movement = createMovement(dMin, '4:0:0', 4);

		expect(movement).toMatchSnapshot();
	});

	it('should generate a basic movement in 3/4', () => {
		// given
		Random.setSeed('test');

		const dMin = new Key('D', Key.Minor);
		// when
		const movement = createMovement(dMin, new Time('2:0:0', [ 3, 4 ]), 6);

		expect(movement).toMatchSnapshot();
	});

	it('should generate a simple movement - euclidean', () => {
		// given
		Random.setSeed('test-2');

		const aMaj = new Key('A', Key.Major);

		const opts = {
			modProb: 0.40,
		};

		// when
		const movement = createMovement(aMaj, '4:0:0', 6, opts);

		expect(movement).toMatchSnapshot();
	});

	it('should generate a simple movement - turn rhythm', () => {
		// given
		Random.setSeed('test');

		const aMaj = new Key('A', Key.Major);

		const opts = {
			modProb: 0.40,
			rhythm: MovementRhythm.Turn,
		};

		// when
		const movement = createMovement(aMaj, '4:0:0', 6, opts);

		expect(movement).toMatchSnapshot();
	});

	it('should generate a long movement - turn rhythm', () => {
		// given
		Random.setSeed('test');

		const aMaj = new Key('A', Key.Major);

		const opts = {
			modProb: 0.40,
			rhythm: MovementRhythm.Turn,
		};

		// when
		const movement = createMovement(aMaj, '8:0:0', 4, opts);

		expect(movement).toMatchSnapshot();
	});
});
