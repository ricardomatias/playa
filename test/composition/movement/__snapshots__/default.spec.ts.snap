// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A Movement test suite should generate a basic movement 1`] = `
{
  "events": [
    {
      "key": "D",
      "position": "VI",
      "root": "D",
      "scaleName": "Aeolian",
      "type": "mode",
    },
    {
      "key": "D",
      "position": "II",
      "root": "G",
      "scaleName": "Dorian",
      "type": "mode",
    },
    {
      "key": "D",
      "position": "V",
      "root": "C",
      "scaleName": "Mixolydian",
      "type": "mode",
    },
    {
      "key": "D",
      "position": "IV",
      "root": "Bb",
      "scaleName": "Lydian",
      "type": "mode",
    },
  ],
  "timeline": [
    {
      "dur": 1920,
      "key": {
        "root": "D",
        "scale": "1P 2M 3m 4P 5P 6m 7m",
      },
      "time": 0,
    },
    {
      "dur": 1920,
      "key": {
        "root": "G",
        "scale": "1P 2M 3m 4P 5P 6M 7m",
      },
      "time": 1920,
    },
    {
      "dur": 1920,
      "key": {
        "root": "C",
        "scale": "1P 2M 3M 4P 5P 6M 7m",
      },
      "time": 3840,
    },
    {
      "dur": 1920,
      "key": {
        "root": "Bb",
        "scale": "1P 2M 3M 4A 5P 6M 7M",
      },
      "time": 5760,
    },
  ],
}
`;

exports[`A Movement test suite should generate a basic movement in 3/4 1`] = `
{
  "events": [
    {
      "key": "D",
      "position": "VI",
      "root": "D",
      "scaleName": "Aeolian",
      "type": "mode",
    },
    {
      "key": "D",
      "position": "II",
      "root": "G",
      "scaleName": "Dorian",
      "type": "mode",
    },
    {
      "key": "D",
      "position": "V",
      "root": "C",
      "scaleName": "Mixolydian",
      "type": "mode",
    },
    {
      "key": "D",
      "position": "IV",
      "root": "Bb",
      "scaleName": "Lydian",
      "type": "mode",
    },
    {
      "key": "D",
      "position": "III",
      "root": "A",
      "scaleName": "Phrygian",
      "type": "mode",
    },
    {
      "key": "D",
      "position": "VI",
      "root": "D",
      "scaleName": "Aeolian",
      "type": "mode",
    },
  ],
  "timeline": [
    {
      "dur": 480,
      "key": {
        "root": "D",
        "scale": "1P 2M 3m 4P 5P 6m 7m",
      },
      "time": 0,
    },
    {
      "dur": 480,
      "key": {
        "root": "G",
        "scale": "1P 2M 3m 4P 5P 6M 7m",
      },
      "time": 480,
    },
    {
      "dur": 480,
      "key": {
        "root": "C",
        "scale": "1P 2M 3M 4P 5P 6M 7m",
      },
      "time": 960,
    },
    {
      "dur": 480,
      "key": {
        "root": "Bb",
        "scale": "1P 2M 3M 4A 5P 6M 7M",
      },
      "time": 1440,
    },
    {
      "dur": 480,
      "key": {
        "root": "A",
        "scale": "1P 2m 3m 4P 5P 6m 7m",
      },
      "time": 1920,
    },
    {
      "dur": 480,
      "key": {
        "root": "D",
        "scale": "1P 2M 3m 4P 5P 6m 7m",
      },
      "time": 2400,
    },
  ],
}
`;

exports[`A Movement test suite should generate a long movement - turn rhythm 1`] = `
{
  "events": [
    {
      "key": "A",
      "position": "I",
      "root": "A",
      "scaleName": "Ionian",
      "type": "mode",
    },
    {
      "key": "F#",
      "position": "I",
      "root": "F#",
      "scaleName": "Ionian",
      "type": "key",
    },
    {
      "key": "F#",
      "position": "IV",
      "root": "B",
      "scaleName": "Lydian",
      "type": "mode",
    },
    {
      "key": "F#",
      "position": "III",
      "root": "A#",
      "scaleName": "Phrygian",
      "type": "mode",
    },
  ],
  "timeline": [
    {
      "dur": 3840,
      "key": {
        "root": "A",
        "scale": "1P 2M 3M 4P 5P 6M 7M",
      },
      "time": 0,
    },
    {
      "dur": 3840,
      "key": {
        "root": "F#",
        "scale": "1P 2M 3M 4P 5P 6M 7M",
      },
      "time": 3840,
    },
    {
      "dur": 3840,
      "key": {
        "root": "B",
        "scale": "1P 2M 3M 4A 5P 6M 7M",
      },
      "time": 7680,
    },
    {
      "dur": 3840,
      "key": {
        "root": "A#",
        "scale": "1P 2m 3m 4P 5P 6m 7m",
      },
      "time": 11520,
    },
  ],
}
`;

exports[`A Movement test suite should generate a simple movement - euclidean 1`] = `
{
  "events": [
    {
      "key": "A",
      "position": "I",
      "root": "A",
      "scaleName": "Ionian",
      "type": "mode",
    },
    {
      "key": "A",
      "position": "II",
      "root": "B",
      "scaleName": "Dorian",
      "type": "mode",
    },
    {
      "key": "A",
      "position": "V",
      "root": "E",
      "scaleName": "Mixolydian",
      "type": "mode",
    },
    {
      "key": "B",
      "position": "V",
      "root": "B",
      "scaleName": "Mixolydian",
      "type": "key",
    },
    {
      "key": "B",
      "position": "IV",
      "root": "A",
      "scaleName": "Lydian",
      "type": "mode",
    },
    {
      "key": "B",
      "position": "IV",
      "root": "A",
      "scaleName": "Lydian",
      "type": "mode",
    },
  ],
  "timeline": [
    {
      "dur": 1440,
      "key": {
        "root": "A",
        "scale": "1P 2M 3M 4P 5P 6M 7M",
      },
      "time": 0,
    },
    {
      "dur": 1440,
      "key": {
        "root": "B",
        "scale": "1P 2M 3m 4P 5P 6M 7m",
      },
      "time": 1440,
    },
    {
      "dur": 960,
      "key": {
        "root": "E",
        "scale": "1P 2M 3M 4P 5P 6M 7m",
      },
      "time": 2880,
    },
    {
      "dur": 1440,
      "key": {
        "root": "B",
        "scale": "1P 2M 3M 4P 5P 6M 7m",
      },
      "time": 3840,
    },
    {
      "dur": 1440,
      "key": {
        "root": "A",
        "scale": "1P 2M 3M 4A 5P 6M 7M",
      },
      "time": 5280,
    },
    {
      "dur": 960,
      "key": {
        "root": "A",
        "scale": "1P 2M 3M 4A 5P 6M 7M",
      },
      "time": 6720,
    },
  ],
}
`;

exports[`A Movement test suite should generate a simple movement - turn rhythm 1`] = `
{
  "events": [
    {
      "key": "A",
      "position": "I",
      "root": "A",
      "scaleName": "Ionian",
      "type": "mode",
    },
    {
      "key": "A",
      "position": "IV",
      "root": "D",
      "scaleName": "Lydian",
      "type": "mode",
    },
    {
      "key": "A",
      "position": "IV",
      "root": "D",
      "scaleName": "Lydian",
      "type": "mode",
    },
    {
      "key": "A",
      "position": "IV",
      "root": "D",
      "scaleName": "Lydian",
      "type": "mode",
    },
    {
      "key": "A",
      "position": "II",
      "root": "B",
      "scaleName": "Dorian",
      "type": "mode",
    },
    {
      "key": "A",
      "position": "III",
      "root": "C#",
      "scaleName": "Phrygian",
      "type": "mode",
    },
  ],
  "timeline": [
    {
      "dur": 160,
      "key": {
        "root": "A",
        "scale": "1P 2M 3M 4P 5P 6M 7M",
      },
      "time": 0,
    },
    {
      "dur": 80,
      "key": {
        "root": "D",
        "scale": "1P 2M 3M 4A 5P 6M 7M",
      },
      "time": 160,
    },
    {
      "dur": 1920,
      "key": {
        "root": "D",
        "scale": "1P 2M 3M 4A 5P 6M 7M",
      },
      "time": 240,
    },
    {
      "dur": 480,
      "key": {
        "root": "D",
        "scale": "1P 2M 3M 4A 5P 6M 7M",
      },
      "time": 2160,
    },
    {
      "dur": 2880,
      "key": {
        "root": "B",
        "scale": "1P 2M 3m 4P 5P 6M 7m",
      },
      "time": 2640,
    },
    {
      "dur": 2160,
      "key": {
        "root": "C#",
        "scale": "1P 2m 3m 4P 5P 6m 7m",
      },
      "time": 5520,
    },
  ],
}
`;
